tests/unittests/MatrixRotationTest.cpp

```diff
@@ -0,0 +1,676 @@
+    size_t j = nextRandInt(i, numSamples - 1);
+    std::swap(rotationRelationshipData[i], rotationRelationshipData[j]);
+  }
+
+  // Convert words and hyphens to a tensor representation.
+  Tensor matricesA(ElemKind::FloatTy, {numSamples, 9});
+  Tensor matricesB(ElemKind::FloatTy, {numSamples, 9});
+  Tensor expected(ElemKind::IndexTy, {numSamples, 1});
+  auto matricesAHandle = matricesA.getHandle<Representation>();
+  auto matricesBHandle = matricesB.getHandle<Representation>();
+  auto expectedHandle = expected.getHandle<size_t>();
+  for (size_t i = 0; i != numSamples; i++) {
+    const RotationRelationship &rel = rotationRelationshipData[i];
+
+    // Check that the data makes sense.
+    ASSERT_LT(rel.matrixIdxA_, numMatrices);
+    ASSERT_LT(rel.matrixIdxB_, numMatrices);
+
+    const Element(*matrixA)[3][3] = &trainingData[rel.matrixIdxA_];
+    const Element(*matrixB)[3][3] = &trainingData[rel.matrixIdxB_];
+
+    Matrix3x3<Element> MA(*matrixA);
+    Matrix3x3<Element> MB(*matrixB);
+    ASSERT_EQ(MA.clockwiseRotationStepsToGet(MB),
+              rel.clockwiseRotationStepsAToB_);
+
+    mapMatrix(*matrixA, i, matricesAHandle);
+    mapMatrix(*matrixB, i, matricesBHandle);
+    expectedHandle.at({i, 0}) = rel.clockwiseRotationStepsAToB_ != -1;
+  }
+
+  // Now build the network.
+  ExecutionEngine EE(BackendKind::CPU);
+  EE.getConfig().learningRate = learningRate;
+  EE.getConfig().batchSize = batchSize;
+  MatrixRotationRecognitionNetwork net(EE.getModule(), EE.getConfig());
+  // Train using mini-batch SGD.
+  EE.compile(CompilationMode::Train, net.train_);
+
+  EE.runBatch(iterations, {net.matrixA_, net.matrixB_, net.expected_},
+              {&matricesA, &matricesB, &expected});
+
+  // Now test inference on the trained network.
+  unsigned cpuErrors = net.inferenceErrors(EE, "cpu", matricesA, matricesB,
+                                           rotationRelationshipData.get());
+  EXPECT_LE(cpuErrors, errors);
+
+  // See of the interpreter gets the same result.
+  EE.setBackend(BackendKind::Interpreter);
+  EXPECT_EQ(net.inferenceErrors(EE, "interpreter", matricesA, matricesB,
+                                rotationRelationshipData.get()),
+            cpuErrors);
+}
+
+TEST(MatrixRotationTest, networkFloat) {
+  const size_t numSamples = sizeof(trainingDataForFloatRelationshipData) /
+                            sizeof(trainingDataForFloatRelationshipData[0]);
+
+  const size_t numMatrices =
+      sizeof(trainingDataForFloat) / sizeof(trainingDataForFloat[0]);
+
+  // The purpose of these asserts is to point out if the training data has been
+  // changed.
+  ASSERT_EQ(numSamples, 3191);
+  ASSERT_EQ(numMatrices, 1594);
+  testNetwork<float>(trainingDataForFloat, numMatrices,
+                     trainingDataForFloatRelationshipData, numSamples,
+                     /*learningRate*/ 0.02, /*batchSize*/ 17,
+                     /*iterations*/ 27000, /*errors*/ 0);
+}
+
+TEST(MatrixRotationTest, networkInt8) {
```

Comment:
"networkInt8" what is the goal here?

The network itself is not the quantized network. I do not think this test brings a lot of value, but makes a lot of things to be templatized (which might not be necessary).

Reply:
Agreed.

That's what I noted in some earlier post:
> Note: I could completely get rid of the templatization of the Matrix3x3 struct and such. The reason I added it in the first place was because I wanted to have a network with FloatTy and one with Int8Ty.
> I realized after the fact that Int8Ty was Int8QTy and cannot be used directly (i.e., it needs to go through quantization).

Reply:
nice!